gnovm/pkg/gnolang/preprocess.go

```diff
@@ -438,8 +438,12 @@ func Preprocess(store Store, ctx BlockNode, n Node) Node {
 					// skip declarations already predefined
 					// (e.g. through recursion for a dependent)
 				} else {
+					d := n.(Decl)
+					if cd, ok := d.(*ValueDecl); ok {
+						checkAssignmentMismatch(cd)
+					}
```

Comment:
This should also be applied in AssignStmt -> DEFINE, though note that there is already a check in type_check.go (see AssignStmt.AssertCompatible); so maybe merge them instead of doing the check twice.

Reply:
Sorry, I didn't understand. In my case, I think I am  checking the ValueDecl (when using var) then in  the type_check.go we are checking the AssignStmt (when using `:=`).

Reply:
yes. but since your error message is good, it should show up also when we have a mismatch with := (ie. `a, b := 1`). Currently, it doesn't show up

Reply:
I just managed the value declaration part for this PR.  I will handle assignment case also if you want

Reply:
Done here: https://github.com/gnolang/gno/pull/2206/commits/3ea7a53397a861f7acb83bdc0dbcd9c10b750143

Should fix this too: https://github.com/gnolang/gno/issues/2637